File: dotnet/src/reader/mxGraphViewImageReader.cs
// Copyright (c) 2007-2008, Gaudenz Alder
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Xml;

namespace com.mxgraph
{
    /// <summary>
    /// A converter that renders display XML data onto a GDI canvas.
    /// </summary>
    public class mxGraphViewImageReader : mxGraphViewReader
    {

        /// <summary>
        /// Specifies the background color.
        /// </summary>
        protected Color? background;

        /// <summary>
        /// Specifies the border size. Default is 0.
        /// </summary>
        protected int border;

        /// <summary>
        /// Default is true.
        /// </summary>
        protected bool antiAlias;

        /// <summary>
        /// Default is true.
        /// </summary>
        protected bool cropping;

        /// <summary>
        /// Specifies the optional clipping rectangle.
        /// </summary>
        protected mxRectangle clip;

        /// <summary>
        /// Constructs a new GDI reader for the given display XML reader.
        /// </summary>
        /// <param name="reader"></param>
        public mxGraphViewImageReader(XmlReader reader) : this(reader, null) { }

        /// <summary>
        /// Constructs a new GDI reader for the given display XML reader.
        /// </summary>
        public mxGraphViewImageReader(XmlReader reader, Color? background) : this(reader, background, 0) { }

        /// <summary>
        /// Constructs a new GDI reader for the given display XML reader.
        /// </summary>
        public mxGraphViewImageReader(XmlReader reader, Color? background, int border)
            : this(reader, background, border, true) { }

        /// <summary>
        /// Constructs a new GDI reader for the given display XML reader.
        /// </summary>
        public mxGraphViewImageReader(XmlReader reader, Color? background, int border, bool antiAlias)
            : this(reader, background, border, antiAlias, true) { }

        /// <summary>
        /// Constructs a new GDI reader for the given display XML reader.
        /// </summary>
        public mxGraphViewImageReader(XmlReader reader, Color? background, int border, bool antiAlias, bool cropping)
        {
            Background = background;
            Border = border;
            AntiAlias = antiAlias;
            Cropping = cropping;
            Read(reader);
        }

        /// <summary>
        /// Accessors for the background property.
        /// </summary>
        public Color? Background
        {
            get { return background; }
            set { background = value; }
        }

        /// <summary>
        /// Accessors for the border property.
        /// </summary>
        public int Border
        {
            get { return border; }
            set { border = value; }
        }

        /// <summary>
        /// Accessors for the background property.
        /// </summary>
        public bool AntiAlias
        {
            get { return antiAlias; }
            set { antiAlias = value; }
        }

        /// <summary>
        /// Accessors for the cropping property.
        /// </summary>
        public bool Cropping
        {
            get { return cropping; }
            set { cropping = value; }
        }

        /// <summary>
        /// Accessors for the clip property.
        /// </summary>
        public mxRectangle Clip
        {
            get { return clip; }
            set { clip = value; }
        }

        /* (non-Dotnetdoc)
         * see com.mxgraph.mxGraphViewReader.CreateCanvas()
         */
        override public mxICanvas CreateCanvas(Dictionary<string, Object> attrs)
        {
            int width = 0;
            int height = 0;
            int dx = 0;
            int dy = 0;

            mxRectangle tmp = Clip;

            if (tmp != null)
            {
                dx -= (int)tmp.X;
                dy -= (int)tmp.Y;
                width = (int)tmp.Width;
                height = (int)tmp.Height;
            }
            else
            {
                int x = (int)Math.Round(mxUtils.GetDouble(attrs, "x"));
                int y = (int)Math.Round(mxUtils.GetDouble(attrs, "y"));
                width = (int)(Math.Round(mxUtils.GetDouble(attrs, "width")))
                    + border + 3;
                height = (int)(Math.Round(mxUtils.GetDouble(attrs, "height")))
                    + border + 3;

                if (cropping)
                {
                    dx = -x + 3;
                    dy = -y + 3;
                }
                else
                {
                    width += x;
                    height += y;
                }
            }

            mxImageCanvas canvas = new mxImageCanvas(new mxGdiCanvas(),
                    width, height, Background, AntiAlias);
            canvas.Translate = new Point(dx, dy);

            return canvas;
        }

        /// <summary>
        /// Creates the image for the given display XML reader. For a given XmlReader,
        /// use the following code to create the view reader:
        /// new mxGraphViewImageReader(xmlReader, background, border, antiAlias);
        /// </summary>
        /// <param name="viewReader">Reader that contains the display XML.</param>
        /// <returns>Returns an image representing the display XML reader.</returns>
        public static Image Convert(mxGraphViewImageReader viewReader)
        {
            if (viewReader.Canvas is mxImageCanvas)
            {
                return ((mxImageCanvas) viewReader.Canvas).Destroy();
            }

            return null;
        }

    }

}

File: php/src/reader/mxGraphViewImageReader.php
<?php
/**
 * Copyright (c) 2006-2013, Gaudenz Alder
 */

class mxGraphViewImageReader
{

	/**
	 * Class: mxGraphViewImageReader
	 *
	 * A display XML to image converter. This allows to create an image of a graph
	 * without having to parse and create the graph model using the XML file
	 * created for the mxGraphView object in the thin client.
	 *
	 * To create the XML for the mxGraphView on the client:
	 *
	 * (code)
	 * var enc = new mxCodec(mxUtils.createXMLDocument());
	 * var node = enc.encode(editor.graph.view);
	 * var xml = mxUtils.getXML(node);
	 * (end)
	 * 
	 * Variable: canvas
	 *
	 * Holds the canvas.
	 */
	var $canvas;
	
	/**
	 * Variable: scale
	 * 
	 * Holds the global scale of the graph. This is set just before
	 * createCanvas is called.
	 */
	var $scale = 1;
	
	/**
	 * Variable: parser
	 *
	 * Holds the SAX parser.
	 */
	var $parser;
	
	/**
	 * Variable: background
	 *
	 * Holds the background color.
	 */
	var $background;
	
	/**
	 * Variable: border
	 *
	 * Holds the border size. Default is 0.
	 */
	var $border;

	/**
	 * Constructor: mxGraphViewImageReader
	 *
	 * Constructs a new image graph view reader.
	 */
	function mxGraphViewImageReader($background = null, $border = 0)
	{
		$this->parser = xml_parser_create();
		
		xml_parser_set_option($this->parser, XML_OPTION_CASE_FOLDING, 0);
		xml_set_object($this->parser, $this);
		xml_set_element_handler($this->parser, "startElement", "endElement");
			
		$this->background = $background;
		$this->border = $border;
	}

	/**
	 * Function: createCanvas
	 *
	 * Returns the canvas to be used for rendering.
	 */
	function createCanvas($attrs)
	{
		$width = $attrs["x"] + $attrs["width"] + $this->border + 1;
		$height = $attrs["y"] + $attrs["height"] + $this->border + 1;

		return new mxGdCanvas($width, $height, $this->scale, $this->background);
	}

	/**
	 * Function: read
	 *
	 * Reads the specified view XML string.
	 */
	function read($string)
	{
		xml_parse($this->parser, $string, true);
	}
	
	/**
	 * Function: readFile
	 *
	 * Reads the specified view XML file in blocks of 4096 bytes.
	 */
	function readFile($filename)
	{
		$fp = fopen($filename, "r");
		
		while ($data = fread($fp, 4096))
		{
			xml_parse($this->parser, $data, feof($fp)) or
			die (sprintf("XML Error: %s at line %d", 
				xml_error_string(xml_get_error_code($this->parser)),
				xml_get_current_line_number($this->parser)));
		}
		
		fclose($fp);
	}

	/**
	 * Function: startElement
	 *
	 * Invoked by the SAX parser when an element starts.
	 */
	function startElement($parser, $name, $attrs)
	{
		if ($this->canvas == null && $name == "graph")
		{
			$this->scale = mxUtils::getValue($attrs, "scale", 1);
			$this->canvas = $this->createCanvas($attrs);
		}
		else if ($this->canvas != null)
		{
			$edge = $name == "edge";
			$group = $name == "group";
			$vertex = $name == "vertex";
			
			if (($edge && isset($attrs["points"])) ||
				(($vertex || $group) && isset($attrs["x"]) &&
				isset($attrs["y"]) && isset($attrs["width"]) &&
				isset($attrs["height"])))
			{
				$state = new mxCellState(null, null, $attrs);

				$label = $this->parseState($state, $edge);
				$this->canvas->drawCell($state);
				$this->canvas->drawLabel($label, $state, false);
			}
		}
	}
	
	/**
	 * Function: parseState
	 *
	 * Parses the bounds, absolute points and label information from the style
	 * of the state into its respective fields and returns the label of the
	 * cell.
	 */
	function parseState($state, $edge)
	{
		$style = $state->style;

		// Parses the bounds
		$state->x = mxUtils::getNumber($style, "x");
		$state->y = mxUtils::getNumber($style, "y");
		$state->width = mxUtils::getNumber($style, "width");
		$state->height = mxUtils::getNumber($style, "height");

		// Parses the absolute points list
		$tmp = mxUtils::getValue($style, "points");
		
		if (strlen($tmp) > 0)
		{
			$pts = $this->parsePoints($tmp);
			
			if (sizeof($pts) > 0)
			{
				$state->absolutePoints = $pts;
			}
		}
		
        // Parses the label and label bounds
		$label = mxUtils::getValue($style, "label");
		
		if ($label != null && strlen($label) > 0)
		{
			$offset = new mxPoint(mxUtils::getNumber($style, "dx"),
				mxUtils::getNumber($style, "dy"));
			$vertexBounds = (!$edge) ? $state : null;
			$state->labelBounds = mxUtils::getLabelPaintBounds($label, $style,
				mxUtils::getValue($style, "html", false), $offset,
				$vertexBounds, $this->scale);
		}
		
		return $label;
	}

	/**
	 * Function: parsePoints
	 * 
	 * Parses a string that represents a list of points into an array of
	 * <mxPoints>.
	 */
	function parsePoints($str)
	{
		$pts = array();
		
		if (isset($str))
		{
		 	$len = strlen($str);
			$tmp = "";
			$x = "";
	
		 	for ($i = 0; $i < $len; $i++)
		 	{
		 		$c = $str[$i];
		 		
		 		if ($c == "," || $c == " ")
		 		{
		 			if (strlen($x) == 0)
		 			{
		 				$x = $tmp;
		 			}
		 			else
		 			{
		 				array_push($pts, new mxPoint($x, $tmp));
		 				$x = "";
		 			}
		 			
	 				$tmp = "";
		 		}
		 		else
		 		{
					$tmp .= $c;	 		
		 		}
		 	}
		 	
		 	array_push($pts, new mxPoint($x, $tmp));
		}
		
	 	return $pts;
	}

	/**
	 * Function: endElement
	 *
	 * Invoked by the SAX parser when an element ends.
	 */
	function endElement($parser, $name)
	{
	 	// ignore
	}

	/**
	 * Destructor: destroy
	 *
	 * Destroy all allocated resources for this reader.
	 */
	function destroy()
	{
	 	$this->canvas->destroy();
	 	xml_parser_free($this->parser);
	}

	/**
	 * Function: convert
	 *
	 * Creates the image for the given display XML string.
	 */
	static function convert($string, $background = null)
	{
		$viewReader = new mxGraphViewImageReader($background);
		$viewReader->read($string);
		$image = $viewReader->canvas->getImage();
		
		return $image;	
	}

	/**
	 * Function: convertFile
	 *
	 * Creates the image for the given display XML file.
	 */
	static function convertFile($filename, $background = null)
	{
		$viewReader = new mxGraphViewImageReader($background);
		$viewReader->readFile($filename);
		$image = $viewReader->canvas->getImage();
		
		return $image;		
	}

}
?>
